### PR TITLE
[26.0] Fix clean jwd task

### PR DESCRIPTION
**Pull request by davelopez**
Fixes the issues detected during the planning here https://github.com/galaxyproject/galaxy/discussions/23107#discussioncomment-17638639

- **`NameError: name 'days' is not defined`** — `days` was read inside nested closures (`get_failed_jobs`, `delete_jwd`) but assigned in the outer function body after the closures were defined, causing an unbound variable error on every run.
- **Query returned `Row` tuples instead of `Job` objects** — `query(model.Job.id)` produced tuples that were passed to `object_store.get_filename()`, which expects a `Job` instance → `AttributeError`.
- **Missing `.all()` on the query** — a bare `Query` is always truthy, so the `if not failed_jobs` early-return check never triggered.
- **Jobs with `object_store_id=None` were excluded** — the `isnot(None)` filter skipped jobs using the default object store from cleanup.
- **No visibility into how many JWDs were deleted** — refactored the core logic into `_cleanup_jwds()` which returns the count of deleted directories, enabling proper testing and logging.

A comprehensive test suite (`TestCleanupJwds`) was added covering all five scenarios above plus edge cases (mixed job states, already-deleted JWDs, etc.).

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/celery/tasks.py [hunk 1]
```diff
@@ -615,18 +615,27 @@ def dispatch_pending_notifications(notification_manager: NotificationManager):
...
+    Returns the number of job working directories deleted.
+    """
 
     def get_failed_jobs():
-        return sa_session.query(model.Job.id).filter(
-            model.Job.state == "error",
-            model.Job.update_time < datetime.datetime.now() - datetime.timedelta(days=days),
-            model.Job.object_store_id.isnot(None),
+        return (
+            sa_session.query(model.Job)
```
Comment (mvdbeek) [inline lib/galaxy/celery/tasks.py]: Materializing the full job is a lot heavier than getting the job ids.
I would skip the `.all()` and instead materialize in the loop which should be lazy (to be confirmed) or add a limit that prevents going OOM when there's a lot of jobs in the timedelta

File: lib/galaxy/celery/tasks.py [hunk 2]
```diff
@@ -637,15 +646,26 @@ def delete_jwd(job):
         except OSError as e:
             log.error(f"Error deleting job working directory: {path} : {e.strerror}")
 
-    days = config.failed_jobs_working_directory_cleanup_days
     failed_jobs = get_failed_jobs()
 
     if not failed_jobs:
         log.info("No failed jobs found within the last %s days", days)
+        return 0
```
Comment (mvdbeek) [inline lib/galaxy/celery/tasks.py]: The query should return an iterator so the counter below could be sufficient ?

File: lib/galaxy/celery/tasks.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/celery/tasks.py]: There's also the gap that there's no recording of what dirs (or time deltas) have been cleaned up already.
If we're recording the job working directory in a followup (suggested un https://github.com/galaxyproject/galaxy/discussions/23107#discussioncomment-17638639) we could reset to null or a specific token so we can ensure we don't iterate infinitely (or until timedelta put the job out of the initial query) over jobs whose working dir has already been cleaned up.

Comment (davelopez) [reply lib/galaxy/celery/tasks.py]: Thanks for the catch Marius!
Now it should be a bit more optimal; it is lazy, and also it should process batches of 100 with `yield_per(100)`. Also changed the syntax to SQLAlchemy 2.0 while at it...

**Comment by davelopez** [comment]
Thanks, Marius, hopefully it's better now :+1:

**Comment by mvdbeek** [comment]
Awesome, yes, this looks good, thank you so much!
